x-pack/plugins/ingest_manager/server/services/epm/archive/index.ts

```diff
@@ -0,0 +1,320 @@
+  if (toplevelDir !== `${manifest.name}-${manifest.version}`) {
+    throw new PackageInvalidArchiveError(
+      `Name ${manifest.name} and version ${manifest.version} do not match top-level directory ${toplevelDir}`
+    );
+  }
+
+  const { name, version, description, type, categories, format_version: formatVersion } = manifest;
+  // check for mandatory fields
+  if (!(name && version && description && type && categories && formatVersion)) {
+    throw new PackageInvalidArchiveError(
+      'Invalid top-level package manifest: one or more fields missing of name, version, description, type, categories, format_version'
+    );
+  }
+
+  const datasets = parseAndVerifyDatasets(paths, name, version);
+  const configTemplates = parseAndVerifyConfigTemplates(manifest);
+
+  return {
+    name,
+    version,
+    description,
+    type,
+    categories,
+    format_version: formatVersion,
+    datasets,
+    config_templates: configTemplates,
+  };
+}
+
+function parseAndVerifyDatasets(paths: string[], pkgName: string, pkgVersion: string): Dataset[] {
+  // A dataset is made up of a subdirectory of name-version/dataset/, containing a manifest.yml
+  let datasetPaths: string[] = [];
+  const datasets: Dataset[] = [];
+
+  // pick all paths matching name-version/dataset/DATASET_PATH/...
+  // from those, pick all unique dataset paths
+  paths
+    .filter((path) => path.startsWith(`${pkgName}-${pkgVersion}/dataset/`))
+    .forEach((path) => {
+      const parts = path.split('/');
+      if (parts.length > 2 && parts[2]) datasetPaths.push(parts[2]);
+    });
+
+  datasetPaths = uniq(datasetPaths);
+
+  datasetPaths.forEach((datasetPath) => {
+    const manifestFile = `${pkgName}-${pkgVersion}/dataset/${datasetPath}/manifest.yml`;
+    const manifestBuffer = cacheGet(manifestFile);
+    if (!paths.includes(manifestFile) || !manifestBuffer) {
+      throw new PackageInvalidArchiveError(
+        `No manifest.yml file found for dataset '${datasetPath}'`
+      );
+    }
+
+    let manifest;
+    try {
+      manifest = yaml.load(manifestBuffer.toString());
+    } catch (error) {
+      throw new PackageInvalidArchiveError(
+        `Could not parse package manifest for dataset '${datasetPath}': ${error}.`
+      );
+    }
+
+    const { title: datasetTitle, release, ingest_pipeline: ingestPipeline, type } = manifest;
+    if (!(datasetTitle && release && type)) {
+      throw new PackageInvalidArchiveError(
+        `Invalid manifest for dataset '${datasetPath}': one or more fields missing of 'title', 'release', 'type'`
+      );
+    }
+    const streams = parseAndVerifyStreams(manifest, datasetPath);
+
+    // default ingest pipeline name see https://github.com/elastic/package-registry/blob/master/util/dataset.go#L26
+    return datasets.push({
+      name: `${pkgName}.${datasetPath}`,
+      title: datasetTitle,
+      release,
+      package: pkgName,
+      ingest_pipeline: ingestPipeline || 'default',
```

Comment:
Seeing this I wonder if we should "enforce" setting ingest_pipeline in packages in the future to simplify our code.

Reply:
Yes, not having optional fields with default values would simplify things a bit. There's a few more places where this happens, and I might even have missed some. I still need to go through the `package-registry` code and check for that.